src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/NoBoxingValueTypes.cs

```diff
@@ -0,0 +1,420 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Linq.Expressions;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Configs;
+using BenchmarkDotNet.Diagnosers;
+using BenchmarkDotNet.Engines;
+using BenchmarkDotNet.Jobs;
+using BenchmarkDotNet.Running;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.ManualTesting.Tests
+{
+    public class NoBoxingValueTypes : IDisposable
+    {
+        private static readonly string _table = DataTestUtility.GetUniqueNameForSqlServer(nameof(NoBoxingValueTypes));
+        private const int _count = 5000;
+        private static readonly ItemToCopy _item;
+        private static readonly IEnumerable<ItemToCopy> _items;
+        private static readonly IDataReader _reader;
+
+        private static readonly string _connString = DataTestUtility.TCPConnectionString;
+
+        private class ItemToCopy
+        {
+            // keeping this data static so the performance of the benchmark is not varied by the data size & shape
+            public int IntColumn { get; } = 123456;
+            public bool BoolColumn { get; } = true;
+        }
+
+        static NoBoxingValueTypes()
+        {
+            _item = new ItemToCopy();
+
+            _items = Enumerable.Range(0, _count).Select(x => _item).ToArray();
+
+            _reader = new EnumerableDataReaderFactoryBuilder<ItemToCopy>(_table)
+                .Add("IntColumn", i => i.IntColumn)
+                .Add("BoolColumn", i => i.BoolColumn)
+                .BuildFactory()
+                .CreateReader(_items)
+            ;
+        }
+
+        public NoBoxingValueTypes()
+        {
+            using (var conn = new SqlConnection(_connString))
+            using (var cmd = conn.CreateCommand())
+            {
+                conn.Open();
+                Helpers.TryExecute(cmd, $@"
+                    CREATE TABLE {_table} (
+                        IntColumn INT NOT NULL,
+                        BoolColumn BIT NOT NULL
+                    )
+                ");
+            }
+        }
+
+        
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup), nameof(DataTestUtility.IsNotAzureServer))]
+        public void Should_Not_Box()
+        { // in debug mode, the double boxing DOES occur, which causes this test to fail (does the JIT "do" less in debug?)
+#if DEBUG
```

Comment:
@DavoudEshtehari are you alright with this?  it's a little hacky, but the JIT does not properly optimize in debug for the test to pass

Reply:
The main unit test goal is to check the correctness of the final code besides increasing code coverage. If you are asking for skipping the debug that would be ok since the release version checks in the product line.
Also, `DataTestUtility.AreConnStringsSetup` doesn't contain Azure connection strings and using  `IsNotAzureServer` is not necessary.
The critical question is that according to the server resources, different frameworks, and future changes in them and our source code, which performance parameters do you want to assess and preserve its validation in the future?

Reply:
This unit test accomplishes that goal -- it uses Benchmark.net to ensure that each value isn't boxed at least once by comparing the actual memory allocated vs expected.

I think the count of rows is high enough that there would not be a feasible amount of code changes that would cause the allocations related to BulkCopy machinery to exceed this threshold.